src/raw_node.rs

```diff
@@ -349,94 +379,259 @@ impl<T: Storage> RawNode<T> {
+        assert!(rd_record.number == rd.number);
+        let raft = &mut self.raft;
+        if rd_record.has_snapshot {
+            raft.raft_log.stable_snap();
+        }
+        if rd_record.last_entry.is_some() {
+            raft.raft_log.stable_entries();
+        }
     }
 
-    /// Grabs the snapshot from the raft if available.
-    #[inline]
-    pub fn snap(&self) -> Option<&Snapshot> {
-        self.raft.snap()
+    fn commit_apply(&mut self, applied: u64) {
+        self.raft.commit_apply(applied);
     }
 
-    /// Advance notifies the RawNode that the application has applied and saved progress in the
-    /// last Ready results.
-    pub fn advance(&mut self, rd: Ready) {
-        self.advance_append(rd);
-        let commit_idx = self.prev_hs.commit;
-        if commit_idx != 0 {
-            // In most cases, prevHardSt and rd.HardState will be the same
-            // because when there are new entries to apply we just sent a
-            // HardState with an updated Commit value. However, on initial
-            // startup the two are different because we don't send a HardState
-            // until something changes, but we do send any un-applied but
-            // committed entries (and previously-committed entries may be
-            // incorporated into the snapshot, even if rd.CommittedEntries is
-            // empty). Therefore we mark all committed entries as applied
-            // whether they were included in rd.HardState or not.
-            self.advance_apply(commit_idx);
+    /// Notifies that the ready of this number has been well persisted.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys with numbers smaller than this one have been persisted.
+    pub fn on_persist_ready(&mut self, number: u64) {
+        let (mut index, mut term) = (0, 0);
+        while let Some(record) = self.records.front() {
+            if record.number > number {
+                break;
+            }
+            let mut record = self.records.pop_front().unwrap();
+
+            if let Some((i, t)) = record.last_entry {
+                index = i;
+                term = t;
+            }
+
+            if !record.messages.is_empty() {
+                self.messages.push(mem::take(&mut record.messages));
+            }
         }
+        if term != 0 {
+            self.raft.on_persist_entries(index, term);
+        }
+    }
+
+    /// Advance notifies the RawNode that the application has applied and saved progress
+    /// in the last Ready results.
+    ///
+    /// Returns the LightReady that contains commit index, committed entries and messages.
+    pub fn advance(&mut self, rd: Ready) -> LightReady {
+        let applied = self.commit_since_index;
+        let light_rd = self.advance_append(rd);
+        self.advance_apply_to(applied);
+        light_rd
+    }
+
+    /// Advance append the ready value synchronously.
```

Comment:
```suggestion
    /// Advances the ready without applying committed entries. `advance_apply` should be used later to update applying progress.
```

Reply:
Appending synchronously does not mean applying synchronously. 
Maybe the "`advance_apply` should be used later to update applying progress." can be removed?

Reply:
The comment doesn't mention whether to apply synchronously, it just suggests the advance is split into two step, and apply is excluded from the first step.

Reply:
Got it.